### PR TITLE
Add single device KD recipe

### PR DESCRIPTION
**Pull request by lindawangg**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
What are the changes made in this PR?
* KD recipe (`knowledge_distillation_single_device.py`) is similar to `lora_finetune_single_device.py`. Main differences are:
  * adds kd loss, currently just `ForwardKLLoss` (in `kd_losses.py`), to CE loss
  * adds teacher model inference to get logits
* KD config: `knowledge_distillation_single_device.yaml`

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

```
CUDA_VISIBLE_DEVICES=0 tune run knowledge_distillation_single_device --config qwen2/knowledge_distillation_single_device
```
##### Llama3.1 KD Training
Legend: KD Llama3.1 student (blue), LoRA Llama3.1 student (orange)
imageimageimage

##### Qwen2 KD Training
Legend: KD Qwen2 0.5B student (green), LoRA Qwen2 0.5B student (grey), LoRA Qwen2 1.5B teacher (blue)
imageimageimage

##### Llama3.1 Eval Results
image

##### Qwen2 Eval Results
image

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

File: recipes/configs/llama3_1/kd_single_device.yaml [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Config for single device knowledge distillation in kd_single_device.py
+# using a teacher and student model
+#
+# This config assumes that you've run the following commands before launching
+# this run:
+#   tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct
+#   tune run lora_finetune_single_device --config llama3_1/8B_lora_single_device
```
Comment (ebsmothers) [inline recipes/configs/llama3_1/kd_single_device.yaml:7]: Can you add a bit more detail here? Tbh when I first looked at it I thought you had accidentally just copy-pasted a `tune run` command from another config 😅 . Maybe just add a couple statements explicitly saying something like "Run this to download the model: {tune download ...}. You will then need to fine-tune the teacher model, you can do this with {tune run...}"

File: recipes/configs/llama3_1/kd_single_device.yaml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    consolidated.00.pth
+  ]
+  recipe_checkpoint: null
+  output_dir: /tmp/kd_model/
+  model_type: LLAMA3
+
+# Teacher checkpoint
+teacher_checkpointer:
+  _component_: torchtune.training.FullModelMetaCheckpointer
+  checkpoint_dir: /tmp/Meta-Llama-3.1-8B-Instruct/lora_finetuned_single_device_epoch_1/
```
Comment (ebsmothers) [inline recipes/configs/llama3_1/kd_single_device.yaml:59]: Should make sure that this directory matches the output of whatever command you give for LoRA finetuning at the top of this file

File: recipes/configs/llama3_1/kd_single_device.yaml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+kd_loss:
+  _component_: torchtune.modules.loss.ForwardKLWithChunkedOutputLoss
+kd_ratio: 0.5
+
+# Training
+epochs: 1
+max_steps_per_epoch: null
+gradient_accumulation_steps: 16
+compile: False
```
Comment (ebsmothers) [inline recipes/configs/llama3_1/kd_single_device.yaml:97]: Out of curiosity, did you try with compile yet?

File: torchtune/modules/loss/kd_losses.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+import torch
+import torch.nn.functional as F
+
+
+class ForwardKLLoss(torch.nn.Module):
+    """
+    The Kullback-Leibler divergence loss for valid indexes.
+    Implementation of https://github.com/jongwooko/distillm/blob/master/distillm/losses.py.
+    """
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:17]: I'm surprised the linter didn't yell at you for this, can we add args (well I guess just single arg) with typehints here?

Comment (ebsmothers) [reply torchtune/modules/loss/kd_losses.py:17]: Also two nits on the link: (1) don't include the period at the end (makes it not clickable), and (2) replace master with a specific commit hash (in case things change in the future)

File: torchtune/modules/loss/kd_losses.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                (batch_size*num_tokens, vocab_size).
+            labels (torch.Tensor): Ground truth labels of shape
+                (batch_size, vocab_size).
+
+        Returns:
+            torch.Tensor: KL divergence loss of shape (1,).
+        """
+
+        teacher_prob = F.softmax(teacher_logits, dim=-1, dtype=torch.float32)
+        inf_mask = torch.isinf(student_logits)
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:43]: Noob q: why would student logits be infinite? Does it mean there is some numerical issue? (I know it's in the original implementation, just curious about the rationale)

File: torchtune/modules/loss/kd_losses.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+class ForwardKLWithChunkedOutputLoss(torch.nn.Module):
+    """
+    Forward KL with chunked outputs that saves memory by only upcasting one chunk at a time.
+
+    Since the model is trained with bf16, before computing KL divergence, we have to upcast
+    it to fp32 for better accuracy and stability. When upcasting happens, the memory usage doubles.
+    Models like llama3 have large vocabulary size and, therefore, have a large output
+    result (bsz, num_tokens, vocab_size). If we chunk on the token level, you can still compute
+    the cross entropy normally, but upcasting only one chunk at a time saves considerable memory.
+    """
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:62]: Similar comment here about init args

File: tests/torchtune/modules/loss/test_kd_losses.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+        # vanilla FKL
+        fkl_loss = ForwardKLLoss(ignore_index=ignore_index)
+        logits = logits.reshape(-1, logits.size(-1))
+        teacher_logits = teacher_logits.reshape(-1, teacher_logits.size(-1))
+        labels = labels.reshape(-1)
+        standard_loss = fkl_loss(logits, teacher_logits, labels)
+
+        # Assert
+        assert_expected(chunked_loss, standard_loss, rtol=1e-2, atol=1e-2)
```
Comment (ebsmothers) [inline tests/torchtune/modules/loss/test_kd_losses.py:56]: Would also be good to run the loss from jongwooko repo with identical sets of values, use that to determine the expected value, then compare both chunked and standard losses to that (that way we know that we have numerical parity with a reference implementation)

File: recipes/configs/llama3_1/kd_single_device.yaml (hunk 3 above)
Comment (lindawangg) [reply recipes/configs/llama3_1/kd_single_device.yaml:97]: Yea tested that compile works (in pink)
image

File: tests/torchtune/modules/loss/test_kd_losses.py (hunk 7 above)
Comment (lindawangg) [reply tests/torchtune/modules/loss/test_kd_losses.py:56]: Good idea. I couldn't find any tests in the repo, so I randomly generated the logits and ran it through the distillm implementation.

File: torchtune/modules/loss/kd_losses.py (hunk 5 above)
Comment (lindawangg) [reply torchtune/modules/loss/kd_losses.py:48]: The teacher_logits could be infinite too. I believe the original implementation only considered that the student logits could be infinite because that's the model that's training. The `inf` in the student logits would cause the torch.sum part to be `inf`.

File: recipes/kd_single_device.py [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+                # For CEWithChunkedOutputLoss, if we compile the entire class
+                # we lose the benefits from the chunked loss.
+                # Therefore, we only compile the cross entropy function + upcasting
+                self._loss_fn.compute_cross_entropy = torch.compile(
+                    self._loss_fn.compute_cross_entropy, backend=backend
+                )
+        else:
+            if self._model_compile:
+                log.info("Compiling loss with torch.compile...")
+                self._loss_fn = torch.compile(self._loss_fn, backend=backend)
```
Comment (ebsmothers) [inline recipes/kd_single_device.py:283]: Oh we have some utilities for this now, you can try to use those instead. See usage in our LoRA single-device recipe [here](https://github.com/pytorch/torchtune/blob/4fbe7b2d4956b3790c51d7a255c0040cf5c38fad/recipes/lora_finetune_single_device.py#L247-L253). (If they don't work for KD out of the box lmk, we can refactor as needed)

File: recipes/kd_single_device.py [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+        self.adapter_params = get_adapter_params(model)
+        self._is_dora = any(["magnitude" in k for k in self.adapter_params.keys()])
+        set_trainable_params(model, self.adapter_params)
+
+        if compile_model:
+            log.info("Compiling model layers with torch.compile...")
+            backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
+            for m in reversed(list(model.modules())):
+                if isinstance(m, modules.transformer.TransformerSelfAttentionLayer):
+                    m.compile(backend=backend)
```
Comment (ebsmothers) [inline recipes/kd_single_device.py:420]: Similar comment here about using the compile utilities

File: recipes/kd_single_device.py [hunk 10]
```diff
@@ -0,0 +1,804 @@
...
+        )
+
+        log.info(f"Model is initialized with precision {self._dtype}.")
+
+        if self._device.type == "cuda":
+            memory_stats = training.get_memory_stats(device=self._device)
+            training.log_memory_stats(memory_stats)
+        return model
+
+    def _setup_teacher_model(
```
Comment (ebsmothers) [inline recipes/kd_single_device.py:463]: Wonder if it's worth it to also compile the teacher model when compile=True?

File: recipes/kd_single_device.py (hunk 8 above)
Comment (ebsmothers) [reply recipes/kd_single_device.py:283]: Can also try compiling KD loss function

File: recipes/kd_single_device.py [hunk 11]
```diff
@@ -0,0 +1,804 @@
...
+                                self._model.parameters(),
+                                max_norm=float(self._clip_grad_norm),
+                            )
+                        self._optimizer.step()
+                        self._optimizer.zero_grad(set_to_none=True)
+                        self._lr_scheduler.step()
+                        # Update the number of steps when the weights are updated
+                        self.global_step += 1
+
+                        loss_to_log = running_loss.item()
```
Comment (ebsmothers) [inline recipes/kd_single_device.py:727]: Nit: I'm not positive, but it might be slightly slower to do things this way. Each `.item()` will cause a sync and you could probably get away with calling just `running_class_loss.item()` and `running_kd_loss.item()` and figuring out `loss_to_log` from that (I think).

**Review by ebsmothers — COMMENTED** [review]
OK I took another, more thorough pass, and this looks great! I left a handful more comments but after that there are no real concerns from me

File: recipes/kd_single_device.py (hunk 10 above)
Comment (lindawangg) [reply recipes/kd_single_device.py:463]: I got an error when trying to compile the teacher model. I'm using torch.no_grad when inferencing the teacher model to save memory consumption. However, it seems that torch.no_grad isn't compatible with torch.compile (https://github.com/pytorch/pytorch/issues/100241).

File: recipes/kd_single_device.py (hunk 11 above)
Comment (lindawangg) [reply recipes/kd_single_device.py:727]: Calling .item() twice instead of 3x helps slightly. I don't see much of a difference: 
image

I removed `running_loss` and added computing from class and kd loss.

File: recipes/configs/llama3_1/kd_single_device.yaml (hunk 3 above)
Comment (lindawangg) [reply recipes/configs/llama3_1/kd_single_device.yaml]: The loss curves between compile and non-compile are fairly similar.
imageimage

I do see some non-determinism during eval. The losses are all around 1.2, but there's slight differences in eval. Also interesting that fine-tuning on alpaca dataset actually hurts performances on all benchmarks but Truthful QA.
image

**Review by joecummings — COMMENTED** [review]
Super nit / discussion point:

kd -> distillation. I think it's much easier to understand at a glance what it is. I didn't understand the abbreviation KD until I read through the PR description.

**Comment by lindawangg** [comment]
> Super nit / discussion point:
> 
> kd -> distillation. I think it's much easier to understand at a glance what it is. I didn't understand the abbreviation KD until I read through the PR description.

We could rename from `kd_single_device` to `knowledge_distillation_single_device`. Distillation might be confusing since there's many types.

**Comment by ebsmothers** [comment]
> We could rename from kd_single_device to knowledge_distillation_single_device. Distillation might be confusing since there's many types.

Yeah this sounds good to me. I agree that it's nice to just be explicit in the recipe name so it's obvious what the recipe is doing (even if the name is a bit longer as a result).

File: tests/recipes/test_kd_single_device.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+            teacher_checkpointer.output_dir={tmpdir} \
+            teacher_checkpointer.model_type={model_type.upper()} \
+            tokenizer._component_=torchtune.models.llama3.llama3_tokenizer \
+            tokenizer.path='{tokenizer_path}' \
+            tokenizer.prompt_template=null \
+            ~tokenizer.merges_file \
+            metric_logger._component_=torchtune.training.metric_logging.DiskLogger \
+            metric_logger.filename={log_file} \
+            compile={compile} \
+            kd_loss._component_=torchtune.modules.loss.ForwardKLWithChunkedOutputLoss \
```
Comment (ebsmothers) [inline tests/recipes/test_kd_single_device.py:91]: Do you actually need this override? This should be the default in the config, right?

Comment (lindawangg) [reply tests/recipes/test_kd_single_device.py]: Removed.

**Review by ebsmothers — APPROVED** [review]
Thank you for adding this!
